openedx/features/qverse_features/registration/signals.py

```diff
@@ -1,20 +1,172 @@
-from __future__ import print_function
-from collections import namedtuple
+import logging
+from csv import DictReader, DictWriter
 
 from django.contrib.auth.models import User
+from django.contrib.auth.hashers import make_password
+from django.db import IntegrityError, transaction
 from django.db.models.signals import post_save
 from django.dispatch import receiver
 
-from openedx.features.qverse_features.registration.models import BulkUserRegistration
+from openedx.features.qverse_features.registration.models import BulkUserRegistration, QVerseUserProfile, Department
+from student.models import UserProfile
 
-UserInfo = namedtuple('UserInfo',
-                      ['username', 'email','first_name', 'last_name',
-                       'other_name', 'mobile_number', 'level',
-                       'programme_and_department']
-                     )
+
+LOGGER = logging.getLogger(__name__)
 
 
 @receiver(post_save, sender=BulkUserRegistration)
-def make_users_from_csv_file(sender, instance, created, **kwargs):
-    # TODO: this would be implemented later once the basic architecture is final.
-    pass
+def create_users_from_csv_file(sender, instance, created, **kwargs):
+    """
+    Creates/Updates users and their profiles.
+
+    Reads data from given csv file having user details. Then creates/updates
+    edx user, edx user profile and qverse user profile accordingly. It also
+    writes the status of creation/update on the given file by adding an
+    extra column named 'status' in it. Also handles the exceptions raised by
+    the functions that it calls.
+
+    Arguments:
+        sender - (ModelBase) responsible to initiate the signal
```

Comment:
```suggestion
        sender (ModelBase): responsible to initiate the signal
```

Please follow edX recommended documentation styling.

Ref: https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python-guidelines.html

Reply:
done.